helm/dagger/templates/engine-statefulset.yaml

```diff
@@ -0,0 +1,108 @@
+        checksum/secret: {{ include (print $.Template.BasePath "/magicache-secret.yaml") . | sha256sum }}
+        {{- end }}
+      {{- end }}
+      labels:
+        name: {{ include "dagger.fullname" . }}-engine
+        {{- include "dagger.labels" . | nindent 8 }}
+    spec:
+      {{- with .Values.engine.tolerations }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .Values.engine.affinity }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      securityContext:
+        runAsUser: 0
+        runAsGroup: 1001
+        fsGroup: 1001
+        fsGroupChangePolicy: "OnRootMismatch"
+      serviceAccountName: {{ include "dagger.serviceAccountName" . }}
+      {{- if .Values.engine.priorityClassName }}
+      priorityClassName: {{ .Values.engine.priorityClassName }}
+      {{- end }}
+      containers:
+        - name: dagger-engine
+          image: {{ if .Values.engine.image.ref }}{{ .Values.engine.image.ref }}{{ else }}registry.dagger.io/engine:v{{ .Chart.Version }}{{ end }}
+          imagePullPolicy: {{ .Values.engine.image.pullPolicy }}
+          {{- if .Values.engine.args }}
+          args:
+            {{- toYaml .Values.engine.args | nindent 12 }}
+          {{- end }}
+          {{- if .Values.magicache.enabled }}
+          env:
+          - name: _EXPERIMENTAL_DAGGER_CACHESERVICE_URL
+            value: {{ required "A magicache url is required" .Values.magicache.url }}
+          envFrom:
+          - secretRef:
+              {{- if .Values.magicache.secretName }}
+              name: {{ .Values.magicache.secretName }}
+              {{- else }}
+              name: {{ include "dagger.fullname" . }}-magicache-token
+              {{- end }}
+          {{- end }}
+          securityContext:
+            privileged: true
+            capabilities:
+              add:
+                - ALL
+          resources: {{- toYaml .Values.engine.resources | nindent 12 }}
+          readinessProbe:
+            exec:
+              command:
+                - sh
+                - -exc
+                - |-
+                  dagger core version
+            {{- if .Values.engine.readinessProbeSettings }}
+            {{- toYaml .Values.engine.readinessProbeSettings | nindent 12 }}
+            {{- end }}
+          volumeMounts:
+            - name: data
+              mountPath: /var/lib/dagger
+            - name: run
+              mountPath: /var/run/buildkit
+            {{- if .Values.engine.config }}
+            - name: config
+              mountPath: /etc/dagger/engine.toml
+              subPath: engine.toml
+            {{- end }}
+      terminationGracePeriodSeconds: {{ .Values.engine.terminationGracePeriodSeconds }}
+      volumes:
+        - name: data
+          hostPath:
+            path: /var/lib/dagger-{{ include "dagger.fullname" . }}
+        - name: run
+          hostPath:
+            path: /var/run/dagger-{{ include "dagger.fullname" . }}
```

Comment:
does the engine play well with 2 processes rw-ing to these concurrently?

Reply:
Not at all.

There should always be a single replica, and each install gets a new name, e.g.:

<img width="1114" alt="image" src="https://github.com/user-attachments/assets/959fe105-2317-400c-8d39-85aa99ff5afd" />

```console
root@d24-10-22:/host/var/lib# ls -lahd dagger*
drwxr-xr-x 4 root root 69 Dec 18 12:08 dagger
drwxr-xr-x 4 root root 69 Dec 18 12:08 dagger-dagger2
drwxr-xr-x 4 root root 69 Dec 18 12:08 dagger-dagger3-dagger-helm
drwxr-xr-x 4 root root 69 Dec 20 17:13 dagger-dagger4-dagger-helm
```

Reply:
ah gotcha, interesting. so the statefulset is not intended to scale horizontally, but via adding additional statefulsets? if I'm understanding correctly, it's probably worth setting `replicas: 1` w/ a comment not to scale